test/e2e/network/topology_hints.go

```diff
@@ -0,0 +1,167 @@
+		e2eskipper.SkipUnlessMultizone(c)
+	})
+
+	ginkgo.It("should distribute endpoints evenly", func() {
+		thLabels := map[string]string{labelKey: clientLabelValue}
+		dsSpec := &appsv1.DaemonSet{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:   "topology-echoserver",
+				Labels: thLabels,
+			},
+			Spec: appsv1.DaemonSetSpec{
+				Selector: &metav1.LabelSelector{
+					MatchLabels: thLabels,
+				},
+				Template: v1.PodTemplateSpec{
+					ObjectMeta: metav1.ObjectMeta{
+						Labels: thLabels,
+					},
+					Spec: v1.PodSpec{
+						Containers: []v1.Container{
+							{
+								Name:  "echoserver",
+								Image: imageutils.GetE2EImage(imageutils.EchoServer),
+								Ports: []v1.ContainerPort{{ContainerPort: 8080}},
+								ReadinessProbe: &v1.Probe{
+									Handler: v1.Handler{
+										HTTPGet: &v1.HTTPGetAction{
+											Port: intstr.FromInt(8080),
+											Path: "/healthz",
+										},
+									},
+									FailureThreshold: 10,
+									PeriodSeconds:    1,
+									SuccessThreshold: 1,
+									TimeoutSeconds:   1,
+								},
+							},
+						},
+					},
+				},
+			},
+			Status: appsv1.DaemonSetStatus{},
+		}
+
+		_, err := c.AppsV1().DaemonSets(f.Namespace.Name).Create(context.TODO(), dsSpec, metav1.CreateOptions{})
+		framework.ExpectNoError(err, "error creating DaemonSet")
+
+		svc := createServiceReportErr(c, f.Namespace.Name, &v1.Service{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: "example-int-port",
+				Annotations: map[string]string{
+					v1.AnnotationTopologyAwareHints: "Auto",
+				},
+			},
+			Spec: v1.ServiceSpec{
+				Selector:                 thLabels,
+				PublishNotReadyAddresses: true,
+				Ports: []v1.ServicePort{{
+					Name:       "example",
+					Port:       80,
+					TargetPort: intstr.FromInt(8080),
+					Protocol:   v1.ProtocolTCP,
+				}},
+			},
+		})
+
+		var expectedPods int32
+		err = wait.Poll(5*time.Second, 3*time.Minute, func() (bool, error) {
+			ds, getErr := c.AppsV1().DaemonSets(f.Namespace.Name).Get(context.TODO(), dsSpec.Name, metav1.GetOptions{})
+			if getErr != nil {
+				return false, getErr
+			}
+			if ds.Status.NumberReady != ds.Status.DesiredNumberScheduled {
+				framework.Logf("Expected %d Pods to be ready, got %d", ds.Status.DesiredNumberScheduled, ds.Status.NumberReady)
+				return false, nil
+			}
+
+			expectedPods = ds.Status.NumberReady
+			return true, nil
+		})
+		framework.ExpectNoError(err, "timed out waiting for DaemonSets to be ready")
+
+		var finalSlices []discoveryv1.EndpointSlice
+		err = wait.Poll(5*time.Second, 3*time.Minute, func() (bool, error) {
+			slices, listErr := c.DiscoveryV1().EndpointSlices(f.Namespace.Name).List(context.TODO(), metav1.ListOptions{LabelSelector: fmt.Sprintf("%s=%s", discoveryv1.LabelServiceName, svc.Name)})
+			if listErr != nil {
+				return false, listErr
+			}
+
+			numEndpoints := 0
+			for _, slice := range slices.Items {
+				numEndpoints += len(slice.Endpoints)
+			}
+			if expectedPods > int32(numEndpoints) {
```

Comment:
The flakes I'm remembering were caused by duplicate EndpointSlices getting created due to an out of date informer cache. Essentially the following sequence.

1. Pod A is created
2. EndpointSlice foo is created targeting Pod A
3. Pod B is created
4. Informer cache doesn't have EndpointSlice foo yet, so it creates EndpointSlice bar targeting Pods A and B

There's since been a number of fixes that should make that sequence much less likely, but I don't think it's completely impossible. The very nature of EndpointSlices means that consumers of the API need to dedup endpoints, so even if there are duplicate endpoints for some period of time, it should not break an implementation. All of that to say, this `>` instead of `!=` was actually intended to cover this potential edge case/flake. Hopefully that logic makes sense.

Reply:
I'm fine if you are fine :smile: